fix: ensure consistent & efficient tokenization

ggdupont opened the pull request:
- replaced call to `sentence.split(" ")` by call to `get_words_from_sentence()`
- added LRU cache for memoization (performance gain to be measured)

Luvata commented:
Nice, since we redo tokenization like three times, adding lru_cache will speedup quite a bit.
I think we also need to verify that `badwords` and `stopwords` will work with the tokenizer, since some badwords are made by more than one word.

e.g:
```python
badwords = ["bad word"]
sentence = "this sentence contains bad word"

words = ModifyingSentences.get_words_from_sentence(sentence, None) # ["this", "sentence", "contains", "bad", "word"]
badwords_ratio = len([word for word in words if word in badwords]) / len(words) # 0
```

HugoLaurencon commented:
sentence.split(" ") and get_words_from_sentence() are two different functions. The last one is used to obtain statistics on the document, so we strip the sentence according to special characters and take the lower characters. This is used for filtering (outputing a Yes/No for keeping the sentence). For sentence.split(“ “), it is used for modifying a document (step before filtering where we remove incorrect words like links for example). You shouldn’t replace sentence.split(“ “) by get_words_from_sentence() because we don’t want to convert to lower case characters during the first step.

Concerning the tokenization, it raises some questions and should be done after filters like the one on special characters, which will not work anymore. So we should wait a little bit to integrate this properly in the code but I plan to do it.

HugoLaurencon commented:
But the cache is a good idea

ggdupont commented:
> sentence.split(" ") and get_words_from_sentence() are two different functions. The last one is used to obtain statistics on the document, so we strip the sentence according to special characters and take the lower characters. This is used for filtering (outputing a Yes/No for keeping the sentence). For sentence.split(“ “), it is used for modifying a document (step before filtering where we remove incorrect words like links for example). You shouldn’t replace sentence.split(“ “) by get_words_from_sentence() because we don’t want to convert to lower case characters during the first step.

Noted. But still `sentence.split(" ")` doesn't work on many non-latin languages (on top of my head: Chinese) so I will suggest to extract this as a "tokenization" method.
 
> Concerning the tokenization, it raises some questions and should be done after filters like the one on special characters, which will not work anymore. So we should wait a little bit to integrate this properly in the code but I plan to do it.

To be clear, I'm not talking about tokenization for the model training, just tokenizing for filtering. These are very likely to be 2 different types of tokenizations since the second one might try to be language agnostic but for filtering that might not be possible.

HugoLaurencon commented:
Thanks for your remarks. Yes for languages without spaces between words, tokenization seems appropriate.

HugoLaurencon commented:
I'm closing this right now because the code for tokenization for each language is not done yet (we should open an issue), but this definitely something that we want to do.
